doc/source/reference/c-api.generalized-ufuncs.rst

```diff
@@ -138,22 +141,35 @@ Notes:
    Each dimension name typically corresponds to one level of looping in the
    elementary function's implementation.
 #. White spaces are ignored.
+#. An integer as a dimension name freezes that dimension to the value,
+#. The name can be suffixed with a question mark, this make the dimension a
+   core dimension only if it exists on the input or output, otherwise 1 is used.
```

Comment:
Should be something like: `otherwise a dimensions of size 1 is inserted into the inputs and squeezed from the outputs.`

Also, we should explain exactly when optional dimensions are used. The rule from matmul is that optional dimensions are used whenever there are enough dimensions for them to be well defined, but matmul doesn't have optional core dimensions that appear on multiple arguments.

For example, how should a signature like `(n,m?),(n,m?)->()` be interpreted if the two inputs are 1d and 2d? There are two reasonable interpretations I can think of:
1. Using the signature `(n),(n)->()`, because not all arguments have enough dimensions.
2. Raising an error, because the behavior is potentially ambiguous.

Reply:
O dear, what did I do - this discussion really belongs in @mattip's PR... Though in his PR I'm not exactly sure what the behaviour is...

Here, I do know: the moment any argument has too few dimensions, flexible dimensions are effectively removed altogether from the ufunc, i.e., it follows your option 1. I think this is most consistent with `matmul`, where, if one were to give an output, one would similarly expect that the dimension would just "not exist".

p.s. And apologies for not adding documentation on this beyond what @mattip had - do look at the tests - I have those for both `?` (flexible) and `|1` (broadcasting) - but still need to add some for numerical axes; so far, for those there are just tests that the signature gets parsed right.

(I had a separate PR where it was per argument, but it became clear to me that was just going to be a mess...)